# Data Preparation for Predicting Employee Attrition 
*By Bhavya Bhargava*<br>

Employee attrition prediction is crucial for organizations aiming to maintain a strong and committed workforce. By leveraging data-driven insights, HR analysts can identify key factors that influence employee departures and take proactive steps to improve retention. Preparing the IBM employee attrition dataset is a critical step in this process, ensuring that the data is clean, structured, and ready for meaningful analysis. Proper data preparation—such as handling missing values, encoding categorical variables, and normalizing numerical features—enhances model accuracy and reliability. By refining the dataset, we can build predictive models that help HR teams make informed decisions, reduce turnover costs, and create a more engaged work environment.

### About the Dataset:
#### Dataset Source:
The IBM HR Employee Attrition dataset is publicly available and is often used for predictive analytics and HR analytics. It contains various employee-related attributes, helping organizations understand factors influencing attrition.

This Dataset was Orginally developed for IBM Watson Studio by Saishruthi Swaminathan, Rich Hagarty but has since been made available to people openly on Kaggle to work on and get insights from.

_Sources:_ <br>
[Visit IBM Developer Page](https://developer.ibm.com/patterns/data-science-life-cycle-in-action-to-solve-employee-attrition-problem/)<br>
[Visit Kaggle Page](https://www.kaggle.com/datasets/pavansubhasht/ibm-hr-analytics-attrition-dataset/data)

_*Let's start with the preparation...*_

**Step 1**: Setting up the environment for Processing and Proper displaying of data


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder
%matplotlib inline

# Set display options for better notebook readability
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)
pd.set_option('display.width', 1000)

# Suppress warnings
import warnings
warnings.filterwarnings('ignore')

**Step 2**: Loading the Dataset for processing

In [2]:
print("Loading the dataset...")
df = pd.read_csv('WA_Fn-UseC_-HR-Employee-Attrition.csv')

Loading the dataset...


**Step 3**: Analyzing the structure of the dataset and the basic information for the attributes

In [4]:
print("\n1. Computing size of dataset:")
print("The shape of data frame:", df.shape)
print("Number of Rows in the dataframe:", len(df))
print("Number of Columns in the dataframe:", len(df.columns))

print("\n2. Column labels in the dataset in column order:")
for column in df.columns:
    print(column)


1. Computing size of dataset:
The shape of data frame: (1470, 35)
Number of Rows in the dataframe: 1470
Number of Columns in the dataframe: 35

2. Column labels in the dataset in column order:
Age
Attrition
BusinessTravel
DailyRate
Department
DistanceFromHome
Education
EducationField
EmployeeCount
EmployeeNumber
EnvironmentSatisfaction
Gender
HourlyRate
JobInvolvement
JobLevel
JobRole
JobSatisfaction
MaritalStatus
MonthlyIncome
MonthlyRate
NumCompaniesWorked
Over18
OverTime
PercentSalaryHike
PerformanceRating
RelationshipSatisfaction
StandardHours
StockOptionLevel
TotalWorkingYears
TrainingTimesLastYear
WorkLifeBalance
YearsAtCompany
YearsInCurrentRole
YearsSinceLastPromotion
YearsWithCurrManager


In [5]:
# Checking the Basic information about the dataset
print("\n3. Generating Basic Information of Attributes:")
print(df.info(verbose=True))


3. Generating Basic Information of Attributes:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470

##### Observations from this:
1. There are 1470 records in the dataset having 36 total attributes.
2. Out of these 26 are Numerical attributes and 9 are Categorical. 
3. None of the attributes seem to be having any Null/ missing values.

To Confirm the third and the last observation.

**Step 4**: Confirming and identifying the Numerical and Categorical Features

In [7]:
numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns
categorical_cols = df.select_dtypes(include=['object']).columns

print("\n4. Numerical Features:")
print(numerical_cols.tolist())
print(len(numerical_cols.to_list()))
print("\nCategorical Features:")
print(categorical_cols.tolist())
print(len(categorical_cols.to_list()))


4. Numerical Features:
['Age', 'DailyRate', 'DistanceFromHome', 'Education', 'EmployeeCount', 'EmployeeNumber', 'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked', 'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']
26

Categorical Features:
['Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus', 'Over18', 'OverTime']
9


##### Observations: 
From this we not only confirmed the number of features of each type but we also identified each one of them.
<br>

**Step 5**: Confirming and Verifying the presence of NULL/ missing values

In [8]:
print("\n5. Checking for Missing Values:")
missing_df = df.isnull().sum().to_frame().rename(columns={0:"Total No. of Missing Values"})
missing_df["% of Missing Values"] = round((missing_df["Total No. of Missing Values"]/len(df))*100,2)
display(missing_df)


5. Checking for Missing Values:


Total No. of Missing Values  % of Missing Values
Age                                                 0                  0.0
Attrition                                           0                  0.0
BusinessTravel                                      0                  0.0
DailyRate                                           0                  0.0
Department                                          0                  0.0
DistanceFromHome                                    0                  0.0
Education                                           0                  0.0
EducationField                                      0                  0.0
EmployeeCount                                       0                  0.0
EmployeeNumber                                      0                  0.0
EnvironmentSatisfaction                             0                  0.0
Gender                                              0                  0.0
HourlyRate                                          0                  0.0
JobInvolvement                                      0                  0.0
JobLevel                                            0                  0.0
JobRole                                             0                  0.0
JobSatisfaction                                     0                  0.0
MaritalStatus                                       0                  0.0
MonthlyIncome                                       0                  0.0
MonthlyRate                                         0                  0.0
NumCompaniesWorked                                  0                  0.0
Over18                                              0                  0.0
OverTime                                            0                  0.0
PercentSalaryHike                                   0                  0.0
PerformanceRating                                   0                  0.0
RelationshipSatisfaction                            0                  0.0
StandardHours                                       0                  0.0
StockOptionLevel                                    0                  0.0
TotalWorkingYears                                   0                  0.0
TrainingTimesLastYear                               0                  0.0
WorkLifeBalance                                     0                  0.0
YearsAtCompany                                      0                  0.0
YearsInCurrentRole                                  0                  0.0
YearsSinceLastPromotion                             0                  0.0
YearsWithCurrManager                                0                  0.0

##### Observations from above:
As confirmed above there are no missing values for any of the attibutes/features of the dataset.

Now let's proceed forward with doing the descriptive analysis of Numerical and Categorical attributes along with checking for unique values and dropping unnecessary features. They will help us have an idea of our data and keep only relevant value for modelling.

**Step 6**:  Descriptive Analysis on Numerical Attributes

In [9]:
print("\n6. Descriptive Analysis on Numerical Attributes:")
display(df[numerical_cols].describe().T)


6. Descriptive Analysis on Numerical Attributes:


count          mean          std     min      25%      50%       75%      max
Age                       1470.0     36.923810     9.135373    18.0    30.00     36.0     43.00     60.0
DailyRate                 1470.0    802.485714   403.509100   102.0   465.00    802.0   1157.00   1499.0
DistanceFromHome          1470.0      9.192517     8.106864     1.0     2.00      7.0     14.00     29.0
Education                 1470.0      2.912925     1.024165     1.0     2.00      3.0      4.00      5.0
EmployeeCount             1470.0      1.000000     0.000000     1.0     1.00      1.0      1.00      1.0
EmployeeNumber            1470.0   1024.865306   602.024335     1.0   491.25   1020.5   1555.75   2068.0
EnvironmentSatisfaction   1470.0      2.721769     1.093082     1.0     2.00      3.0      4.00      4.0
HourlyRate                1470.0     65.891156    20.329428    30.0    48.00     66.0     83.75    100.0
JobInvolvement            1470.0      2.729932     0.711561     1.0     2.00      3.0      3.00      4.0
JobLevel                  1470.0      2.063946     1.106940     1.0     1.00      2.0      3.00      5.0
JobSatisfaction           1470.0      2.728571     1.102846     1.0     2.00      3.0      4.00      4.0
MonthlyIncome             1470.0   6502.931293  4707.956783  1009.0  2911.00   4919.0   8379.00  19999.0
MonthlyRate               1470.0  14313.103401  7117.786044  2094.0  8047.00  14235.5  20461.50  26999.0
NumCompaniesWorked        1470.0      2.693197     2.498009     0.0     1.00      2.0      4.00      9.0
PercentSalaryHike         1470.0     15.209524     3.659938    11.0    12.00     14.0     18.00     25.0
PerformanceRating         1470.0      3.153741     0.360824     3.0     3.00      3.0      3.00      4.0
RelationshipSatisfaction  1470.0      2.712245     1.081209     1.0     2.00      3.0      4.00      4.0
StandardHours             1470.0     80.000000     0.000000    80.0    80.00     80.0     80.00     80.0
StockOptionLevel          1470.0      0.793878     0.852077     0.0     0.00      1.0      1.00      3.0
TotalWorkingYears         1470.0     11.279592     7.780782     0.0     6.00     10.0     15.00     40.0
TrainingTimesLastYear     1470.0      2.799320     1.289271     0.0     2.00      3.0      3.00      6.0
WorkLifeBalance           1470.0      2.761224     0.706476     1.0     2.00      3.0      3.00      4.0
YearsAtCompany            1470.0      7.008163     6.126525     0.0     3.00      5.0      9.00     40.0
YearsInCurrentRole        1470.0      4.229252     3.623137     0.0     2.00      3.0      7.00     18.0
YearsSinceLastPromotion   1470.0      2.187755     3.222430     0.0     0.00      1.0      3.00     15.0
YearsWithCurrManager      1470.0      4.123129     3.568136     0.0     2.00      3.0      7.00     17.0

#### Observations from the above:
From the above we get a general idea of the overall spread and average of the values for the numerical attributes. Here, we have noticed a few unnecessary columns which will only cause confusion or potentially cause inaccuracies in the model, so we will proceed with removing them.

**Step 7**:  Dropping/Removing Unnecessary columns

In [10]:
print("\n7. Identifying and Dropping Unnecessary Columns:")
columns_to_drop = ['EmployeeCount', 'EmployeeNumber', 'Over18', 'StandardHours']
df_processed = df.drop(columns_to_drop, axis=1)
print("Columns dropped:", columns_to_drop)
print("New shape:", df_processed.shape)


7. Identifying and Dropping Unnecessary Columns:
Columns dropped: ['EmployeeCount', 'EmployeeNumber', 'Over18', 'StandardHours']
New shape: (1470, 31)


Now with the removal of unnecessary columns we can proceed with doing the analysis for the categorical columns.

**Step 8**:  Descriptive Analysis on Categorical Attributes

In [11]:
print("\n8. Descriptive Analysis on Categorical Attributes:")
display(df_processed.describe(include='object').T)


8. Descriptive Analysis on Categorical Attributes:


count unique                     top  freq
Attrition       1470      2                      No  1233
BusinessTravel  1470      3           Travel_Rarely  1043
Department      1470      3  Research & Development   961
EducationField  1470      6           Life Sciences   606
Gender          1470      2                    Male   882
JobRole         1470      9         Sales Executive   326
MaritalStatus   1470      3                 Married   673
OverTime        1470      2                      No  1054

##### Observations from the above:
This confirms the overall count of records for each of these attributes. Along with that we can also see the values appearing the most for each of these with their appearence frequency. We also see the count of the unique values for each attribute. But, to be certain and to check which are the unique values we should do a check.

**Step 9**:  Checking unique values in Categorical Attributes

In [12]:
print("\n9. Unique Values in Categorical Attributes:")
for column in df_processed.select_dtypes(include=['object']).columns:
    print(f"\n{column}:")
    print(df_processed[column].value_counts())
    print(f"Number of unique values: {df_processed[column].nunique()}")


9. Unique Values in Categorical Attributes:

Attrition:
Attrition
No     1233
Yes     237
Name: count, dtype: int64
Number of unique values: 2

BusinessTravel:
BusinessTravel
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: count, dtype: int64
Number of unique values: 3

Department:
Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64
Number of unique values: 3

EducationField:
EducationField
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: count, dtype: int64
Number of unique values: 6

Gender:
Gender
Male      882
Female    588
Name: count, dtype: int64
Number of unique values: 2

JobRole:
JobRole
Sales Executive              326
Research Scientist           292
Laboratory Technician        259
Manufacturing Director       145
Healthcare Representative    131
Manager          

This confirms and give us a good Idea of the unique values and their counts. 
<br>
<br>
Now, as we have understood the data we can proceed with the specific processing (including value conversions and relevant feature creation) of this dataset for future Attrition prediction.
<br>
<br>
**Step 10**:  Conducting Necessary data processing for Attrition Prediction

In [14]:
print("\n10. Processing Data for Attrition Prediction...")

# Convert Attrition to binary
df_processed['Attrition'] = (df_processed['Attrition'] == 'Yes').astype(int)

# Replace numeric categories with meaningful labels
category_mappings = {
    "Education": {
        1: "Below College",
        2: "College",
        3: "Bachelor",
        4: "Master",
        5: "Doctor"
    },
    "EnvironmentSatisfaction": {
        1: "Low",
        2: "Medium",
        3: "High",
        4: "Very High"
    },
    "JobInvolvement": {
        1: "Low",
        2: "Medium",
        3: "High",
        4: "Very High"
    },
    "JobSatisfaction": {
        1: "Low",
        2: "Medium",
        3: "High",
        4: "Very High"
    },
    "PerformanceRating": {
        1: "Low",
        2: "Good",
        3: "Excellent",
        4: "Outstanding"
    },
    "RelationshipSatisfaction": {
        1: "Low",
        2: "Medium",
        3: "High",
        4: "Very High"
    },
    "WorkLifeBalance": {
        1: "Bad",
        2: "Good",
        3: "Better",
        4: "Best"
    }
}

for column, mapping in category_mappings.items():
    if column in df_processed.columns:
        df_processed[column] = df_processed[column].replace(mapping)

# Create relevant features for attrition prediction
# Satisfaction score
satisfaction_columns = ['EnvironmentSatisfaction', 'JobSatisfaction', 
                       'RelationshipSatisfaction', 'WorkLifeBalance']
df_processed['OverallSatisfaction'] = df_processed[satisfaction_columns].apply(lambda x: x.map({
    'Low': 1, 'Bad': 1,
    'Medium': 2, 'Good': 2,
    'High': 3, 'Better': 3,
    'Very High': 4, 'Best': 4
}).mean(), axis=1)

# Career growth indicators
df_processed['CareerProgressionRatio'] = (df_processed['YearsSinceLastPromotion'] + 1) / (df_processed['YearsAtCompany'] + 1)
df_processed['YearsWithoutPromotion'] = df_processed['YearsInCurrentRole'] - df_processed['YearsSinceLastPromotion']

# Compensation factors
df_processed['AnnualSalary'] = df_processed['MonthlyIncome'] * 12
df_processed['SalaryHikeImpact'] = df_processed['PercentSalaryHike'] * df_processed['MonthlyIncome']

# Work-life balance indicators
df_processed['WorkLoadIndex'] = df_processed.apply(
    lambda x: 2 if x['OverTime'] == 'Yes' else 1, axis=1) * \
    df_processed['JobInvolvement'].map({'Low': 1, 'Medium': 2, 'High': 3, 'Very High': 4})

# Encode categorical variables
print("\n11. Encoding categorical variables...")
le = LabelEncoder()
categorical_columns = df_processed.select_dtypes(include=['object']).columns
categorical_columns = [col for col in categorical_columns if col != 'Attrition']  # Exclude Attrition as it's already binary

for column in categorical_columns:
    df_processed[f'{column}_Encoded'] = le.fit_transform(df_processed[column])

# View of processed_data for verification
df_processed.head()


10. Processing Data for Attrition Prediction...

11. Encoding categorical variables...


Age  Attrition     BusinessTravel  DailyRate              Department  DistanceFromHome      Education EducationField EnvironmentSatisfaction  Gender  HourlyRate JobInvolvement  JobLevel                JobRole JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  NumCompaniesWorked OverTime  PercentSalaryHike PerformanceRating RelationshipSatisfaction  StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  OverallSatisfaction  CareerProgressionRatio  YearsWithoutPromotion  AnnualSalary  SalaryHikeImpact  WorkLoadIndex  BusinessTravel_Encoded  Department_Encoded  Education_Encoded  EducationField_Encoded  EnvironmentSatisfaction_Encoded  Gender_Encoded  JobInvolvement_Encoded  JobRole_Encoded  JobSatisfaction_Encoded  MaritalStatus_Encoded  OverTime_Encoded  PerformanceRating_Encoded  RelationshipSatisfaction_Encoded  WorkLifeBalance_Encoded
0   41          0      Travel_Rarely       1102                   Sales                 1        College  Life Sciences                  Medium  Female          94           High         2        Sales Executive       Very High        Single           5993        19479                   8      Yes                 11         Excellent                      Low                 0                  8                      0             Bad               6                   4                        0                     5                 2.00                0.142857                      4         71916             65923              6                       2                   2                  2                       1                                2               0                       0                7                        3                      2                 1                          0                                 1                        0
1   49          0  Travel_Frequently        279  Research & Development                 8  Below College  Life Sciences                    High    Male          61         Medium         2     Research Scientist          Medium       Married           5130        24907                   1       No                 23       Outstanding                Very High                 1                 10                      3          Better              10                   7                        1                     7                 3.00                0.181818                      6         61560            117990              2                       1                   1                  1                       1                                0               1                       2                6                        2                      1                 0                          1                                 3                        2
2   37          0      Travel_Rarely       1373  Research & Development                 2        College          Other               Very High    Male          92         Medium         1  Laboratory Technician            High        Single           2090         2396                   6      Yes                 15         Excellent                   Medium                 0                  7                      3          Better               0                   0                        0                     0                 3.00                1.000000                      0         25080             31350              4                       2                   1                  2                       4                                3               1                       2                2                        0                      2                 1                          0                                 2                        2
3   33          0  Travel_Frequently       1392  Research & Development                 3         Master  Life Sciences               Very High  Fe